File: app/src/main/java/good/damn/editor/mediastreaming/camera/avc/cache/MSPacketCombiner.java
package good.damn.editor.mediastreaming.camera.avc.cache;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.HashMap;
import java.util.concurrent.ConcurrentLinkedQueue;

public final class MSPacketCombiner {

    private static final String TAG = "MSPacketCombiner";

    private static final int CACHE_PACKET = 128;

    private final MSPacketFrame[] mPackets =
        new MSPacketFrame[CACHE_PACKET];

    public final void write(
        final int packetId,
        final short chunkId,
        final short chunkCount,
        final byte[] data,
        final MSListenerOnCombinePacket onCombinePacket
    ) {

        @Nullable
        final MSPacketFrame frame = mPackets[
            packetId % CACHE_PACKET
        ];

        if (frame == null) {
            mPackets.put(
                packetId,
                new MSPacketFrame(
                    new MSPacket[
                        chunkCount
                    ],1
                )
            );
            return;
        }

        @NonNull
        final MSPacket[] chunks = frame
            .getChunks();

        if (chunks[chunkId] == null) {
            chunks[chunkId] = new MSPacket(
                data
            );
            frame.setChunkCountAdded(
                frame.getChunkCountAdded() + 1
            );
        }

        if (frame.getChunkCountAdded() >= chunkCount) {
            onCombinePacket.onCombinePacket(
                packetId,
                frame
            );
        }

    }

}

File: app/src/main/java/good/damn/editor/mediastreaming/camera/avc/cache/MSPacketCombiner.kt
package good.damn.editor.mediastreaming.camera.avc.cache

class MSPacketCombiner {

    private val mPackets = HashMap<
        Int, MSPacketFrame
    >()

    fun write(
        packetId: Int,
        chunkId: Short,
        chunkCount: Short,
        data: ByteArray,
        onCombinePacket: MSListenerOnCombinePacket
    ) {
        mPackets[
            packetId
        ]?.apply {

            chunks[
                chunkId
            ] = MSPacket(
                data
            )

            if (chunks.size.toShort() == chunkCount) {
                onCombinePacket.onCombinePacket(
                    packetId,
                    this
                )
                mPackets.remove(
                    packetId
                )
            }

            return
        }

        mPackets[
            packetId
        ] = MSPacketFrame(
            hashMapOf(
                chunkId to MSPacket(
                    data
                )
            )
        )
    }

}